### ISSUE TITLE
Increased base RAM usage and memory leak after upgrade from v23 to v24 in dockerized service

### ISSUE THREAD

**Issue by lnksz** [keycloak/keycloak#28671]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

dist/quarkus

### Describe the bug

After upgrading our dockerized Keycloak service from tag 23.0 to 24.0 (latest) I noticed the following (regressions?).

It is reproducible also with the example from https://www.keycloak.org/server/containers#_writing_your_optimized_keycloak_dockerfile

With H2 for ease of setup and 23.0 as start case:

```
FROM quay.io/keycloak/keycloak:23.0 as builder

# Enable health and metrics support
ENV KC_HEALTH_ENABLED=true
ENV KC_METRICS_ENABLED=true

# Configure a database vendor
# ENV KC_DB=postgres

WORKDIR /opt/keycloak
# for demonstration purposes only, please make sure to use proper certificates in production instead
RUN keytool -genkeypair -storepass password -storetype PKCS12 -keyalg RSA -keysize 2048 -dname "CN=server" -alias server -ext "SAN:c=DNS:localhost,IP:127.0.0.1" -keystore conf/server.keystore
RUN /opt/keycloak/bin/kc.sh build

FROM quay.io/keycloak/keycloak:23.0
COPY --from=builder /opt/keycloak/ /opt/keycloak/

# change these values to point to a running postgres instance
# ENV KC_DB=postgres
# ENV KC_DB_URL=<DBURL>
# ENV KC_DB_USERNAME=<DBUSERNAME>
# ENV KC_DB_PASSWORD=<DBPASSWORD>
ENV KC_HOSTNAME=localhost
ENTRYPOINT ["/opt/keycloak/bin/kc.sh"]
```

## A) Considerable increase in the initial RAM allocation

- 23.0 starts up with ca 550MB on a 32GB system
- 24.0 starts up with ca 1.7GB on the same 32 GB system

Did I miss any hint in https://www.keycloak.org/docs/latest/upgrading/index.html#migrating-to-24-0-0 on why this 3x RAM increase would be the case? 

## B) RAM usage is creeping up over time

In a staging deployment (16GB RAM), with no real load on Keycloak (one extra realm, but no clients/users interacting with the system) I can see how Keycloak v24 is eating up the server's RAM. I included the time series from the v23 container for comparison as well.

![image](https://github.com/keycloak/keycloak/assets/15908859/65d479f4-83aa-4725-88d7-d038a2846f95)

Note: This graph is from a KC container, with more customization, but I will run a container from the above Dockerfile to monitor the RAM usage over time, to see whether the leak is present in the "reference" setup.
 

### Version

24.0.2

### Regression

- [X] The issue is a regression

### Expected behavior

- Keycloak doesn't consume 3x more RAM after 23->24 upgrade
- KC v24 doesn't leak RAM, especially in a nearly idle set-up

### Actual behavior

- 3x more startup RAM consumption
- memory leak

### How to Reproduce?

See Dockerfile example in description taken from https://www.keycloak.org/server/containers#_writing_your_optimized_keycloak_dockerfile

- comment out `ENV KC_DB=postgres` to use H2 DB
- set `23.0` tag for image `mykc:23`
- build with `docker build . -t mykc:23`
- run v23 container with `docker run --rm --name mykc23 -p 8443:8443 -e KEYCLOAK_ADMIN=admin -e KEYCLOAK_ADMIN_PASSWORD=change_me mykc:23 start --optimized` also based on the documentation
- assess RAM usage (e.g. with portainer or with `docker stats mykc23`)

Variant with latest image
- set tag to `latest` or `24.0`
- build with `docker build . -t mykc:24`
- run v24 container with `docker run --rm --name mykc24 -p 8443:8443 -e KEYCLOAK_ADMIN=admin -e KEYCLOAK_ADMIN_PASSWORD=change_me mykc:24 start --optimized`
- assess RAM usage (e.g. with portainer or with `docker stats mykc24`)

**Comment by shawkins** [comment]
Possibly related to #28211 and #28444 

> Did I miss any hint in https://www.keycloak.org/docs/latest/upgrading/index.html#migrating-to-24-0-0 on why this 3x RAM increase would be the case?

See https://github.com/keycloak/keycloak/issues/28211#issuecomment-2020272745

**Comment by lnksz** [comment]
@shawkins thanks for the quick reply! Sorry I didn't see #28211 because I only searched for open tickets...

With `--memory=1G` in the `docker run` command, the initial RAM usage of the the v24 container dropped to ca 500M, which is harmony with `-XX:InitialRAMPercentage=50`. I guess based on `-XX:MaxRAMPercentage=70`, this will go up to 700M, right?

What surprises me, that this RAM eating is taking place even though the KC container has 0 "load". No UI, requests, maintenance, anything...
Is it normal that KC continues to eat up the container memory in an idle state?

I will measure this now with the 1GB limit, because I would expect that an idle service find a (non max) heap size and stays there until some work needs to be done.

**Comment by shawkins** [comment]
> this will go up to 700M, right?

Yes the heap will be allowed to grow up to 700M - the overall java process memory can be higher than that.

> What surprises me, that this RAM eating is taking place even though the KC container has 0 "load". No UI, requests, maintenance, anything...
Is it normal that KC continues to eat up the container memory in an idle state?

I'm not sure what exactly is going on here. I would not expect the heap size to be adjusted higher unless there is pressure, and it doesn't seem like it would be done so smoothly / continuously. But it could be that the particular garbage collector in use will simply allow things to grow to MaxRAMPercentage over time based upon any amount of garbage (even just probes). I have not witnessed this behavior with local testing, so there may be some subtle difference running in Docker that needs another look.

**Comment by lnksz** [comment]
I have run the v24 as from the documentation with 1G limit. The RAM still grows slowly, even though I just let the service run without interacting with it in any way.

![image](https://github.com/keycloak/keycloak/assets/15908859/6fcf4999-3928-4531-83f9-ec6189dab4d7)

I captured the RAM usage via:
```sh
while true; do
	ram="$(docker stats --format=json --no-stream mykc | jq -r '.MemUsage' | cut -d' ' -f1)"
	ts=$(date +%s)
	echo "$ts,$ram" >> kc24-ram.csv
	sleep 30s
done
```

Then visualized it in Excel quickly.

**Comment by sschu** [comment]
@lnksz There are some background tasks running like the ones for expiring old session. The garbage collection can also be quite lazy to reclaim space. So my bet would be it is just not doing collections until it gets close to the maximum which is 70% of the 1G. But you could check by enabling metrics, there should be a counter for the number of small/large garbage collections.

**Comment by lnksz** [comment]
Hi there,

I have run our "real" deployment of the KC configuration and monitored it over the weekend.
( with 1G limit in place)

![image](https://github.com/keycloak/keycloak/assets/15908859/43ddb8fd-fa40-442c-b571-2875b9abbf57)

So after the limit is set, over the long-run there is no such "run away" like with the unlimited version.

In the above screenshot one can see the v23 version (without limit) around 400M, then the unlimited v24 version (saw) and then the 1G limited v24 version.

This is now, what I would expect to see after an upgrade.
Maybe you could add a hint in the v24 upgrade guide? It is still not clear to me, what change exactly caused this in v24...

**Comment by mabartos** [comment]
@lnksz Thanks for the additional info. I just wanted to comment that it should be related only to the limit. 

I created a follow-up issue to improve the docs: https://github.com/keycloak/keycloak/issues/28729
